x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java

```diff
@@ -257,41 +281,109 @@ static Response getOperationsBetween(IndexShard indexShard, long minSeqNo, long
+
+            // TODO: optimize this so that we fetch doc values data in segment and doc id order (see: DocValueFieldsFetchSubPhase):
+            int leafReaderIndex = ReaderUtil.subIndex(scoreDoc.doc, searcher.getIndexReader().leaves());
+            LeafReaderContext leafReaderContext = searcher.getIndexReader().leaves().get(leafReaderIndex);
+            int segmentDocId = scoreDoc.doc - leafReaderContext.docBase;
+            long version = readNumberDvValue(leafReaderContext, VersionFieldMapper.NAME, segmentDocId);
+            long seqNo = readNumberDvValue(leafReaderContext, SeqNoFieldMapper.NAME, segmentDocId);
+            long primaryTerm = readNumberDvValue(leafReaderContext, SeqNoFieldMapper.PRIMARY_TERM_NAME, segmentDocId);
+
+            // TODO: handle NOOPs
+            // NOTE: versionType can always be INTERNAL,
+            //       because version logic has already been taken care of when indexing into leader shard.
+            final VersionType versionType = VersionType.INTERNAL;
+            final Translog.Operation op;
+            if (isDeleteOperation(leafReaderContext, segmentDocId)) {
+                BytesRef idBytes = Uid.encodeId(id);
+                Term uidForDelete = new Term(IdFieldMapper.NAME, idBytes);
+                op = new Translog.Delete(type, id, uidForDelete, seqNo, primaryTerm, version, versionType);
+            } else {
+                // NOTE: autoGeneratedIdTimestamp can always be -1,
+                //       because auto id generation has already been performed when inxdexing into leader shard.
+                final int autoGeneratedId = -1;
+                op = new Translog.Index(type, id, seqNo, primaryTerm, version, versionType, source, routing, autoGeneratedId);
+            }
+            seenBytes += op.estimateSize();
+            operations.add(op);
+            if (seenBytes > byteLimit) {
+                return operations;
             }
         }
+        return operations;
+    }
 
-        if (nextExpectedSeqNo >= maxSeqNo) {
-            return new Response(operations.toArray(EMPTY_OPERATIONS_ARRAY));
-        } else {
-            String message = "Not all operations between min_seq_no [" + minSeqNo + "] and max_seq_no [" + maxSeqNo +
-                    "] found, tracker checkpoint [" + nextExpectedSeqNo + "]";
-            throw new IllegalStateException(message);
+    private static long readNumberDvValue(LeafReaderContext leafReaderContext, String fieldName, int segmentDocId) throws IOException {
+        NumericDocValues versionDvField = leafReaderContext.reader().getNumericDocValues(fieldName);
+        assert versionDvField != null : fieldName + " field is missing";
+        boolean advanced = versionDvField.advanceExact(segmentDocId);
+        assert advanced;
+        return versionDvField.longValue();
+    }
+
+    private static boolean isDeleteOperation(LeafReaderContext leafReaderContext, int segmentDocId) throws IOException {
```

Comment:
this is not correct I think we are blocked here unitl we can identify tombstones